idlpy/src/types.c

```diff
@@ -25,7 +25,32 @@
 #include "naming.h"
 #include "types.h"
 
+//Reserved Python keywords support (Issue 105):
+static const char *python_keywords[] = {
+#define _py_(str) "_" str
+    _py_("False"), _py_("None"), _py_("True"), _py_("and"), _py_("as"), _py_("assert"), 
+    _py_("break"), _py_("class"), _py_("continue"), _py_("def"), _py_("del"), _py_("elif"), 
+    _py_("else"), _py_("except"), _py_("finally"), _py_("for"), _py_("from"), _py_("global"), 
+    _py_("if"), _py_("import"), _py_("in"), _py_("is"), _py_("lambda"), _py_("nonlocal"), 
+    _py_("not"), _py_("or"), _py_("pass"), _py_("raise"), _py_("return"), _py_("try"), 
+    _py_("while"), _py_("with"), _py_("yield")
+#undef _py_
+};
+
+const char *get_c_name(const char *name)
```

Comment:
I think `get_c_name` is not very descriptive to what it actually does: escape python keywords. I think it is actually prudent to define `idlpy_identifier` that calls `idl_identifier` and then filters keywords. This simplifies your code in quite a few places since there is no modified vs non-modified handling needed.

There are a couple places where a name of something that is not a node is determined. For that you can define a `filter_python_keywords` function and use that inside `idlpy_identifier`.

There is a few non-keyword things you should also add to the filter, namely `idl`, `annotate`, `types`, `auto`, `TYPE_CHECKING`, `Optional` and `dataclass` since they are occuring as imports at the top of any idl python output. See also https://github.com/eclipse-cyclonedds/cyclonedds-python/issues/161.

Reply:
Ok, so, we understand you don't want us to follow the cxx way of implementation. 

To confirm with you:

1. Modify name of function `get_c_name `to `filter_python_keywords `
2. `filter_python_keywords` never returns NULL, returns the original name or the original name with the prefix "_".
3. We implement a new function `const char *idlpy_identifier(const void *node)` which internally is going to call:

```
   name = idl_identifier(node) 
   name = filter_python_keywords(name)
   return  name;
```

About more keywords, OK, we will also review this #161 to add all the required keywords. We have filtered the Python keywords indicated in this web page:

https://www.w3schools.com/python/python_ref_keywords.asp
https://www.geeksforgeeks.org/python-keywords/

But, only for curiosity, why do we have to add other keywords that are not considered reserved?, We suppose that the Python code generated doesn't work if it contains any of the "extra" keywords, does it?

Reply:
The proposed modifications sound okay to me.

As for your questions, indeed, they are not keywords but are used in the generated code:

```python
from enum import auto
from typing import TYPE_CHECKING, Optional
from dataclasses import dataclass

import cyclonedds.idl as idl
import cyclonedds.idl.annotations as annotate
import cyclonedds.idl.types as types
```